pkg/partition/consistent/hashing.go

```diff
@@ -33,58 +41,145 @@ func (th rkvHash) Hash(key []byte) uint64 {
-	localNodes := []rkvNode{}
-	azs := h.AzHashing.LocateNodes(key, h.LocalCount)
-	if len(azs) != h.LocalCount {
-		return nil, nil, fmt.Errorf("failed to get %d zones. The return number is %d", h.LocalCount, len(azs))
+func (sahm SyncByZoneAsyncHashingManager) GetSyncNodes(key []byte) ([]Node, error) {
+	localNodes := make([]Node, 0)
+	if sahm.LocalCount < 1 {
+		return localNodes, nil
+	}
+	nodesWithLatency := make([]RkvNode, 0)
+	azs := sahm.AzHashing.LocateNodes(key, sahm.LocalCount)
+	if len(azs) != sahm.LocalCount {
+		return nil, fmt.Errorf("failed to get %d zones. The return number is %d", sahm.LocalCount, len(azs))
 	}
 	for _, az := range azs {
-		lnodes := h.LocalHashing[constants.AvailabilityZone(az.String())].LocateNodes(key, 1)
+		lnodes := sahm.LocalHashing[constants.AvailabilityZone(az.String())].LocateNodes(key, 1)
 		if len(lnodes) != 1 {
-			return nil, nil, fmt.Errorf("failed to get 1 local node. The return number is %d", len(lnodes))
+			return nil, fmt.Errorf("failed to get 1 local node. The return number is %d", len(lnodes))
 		}
-		localNodes = append(localNodes, rkvNode(lnodes[0].String()))
+		nodesWithLatency = append(nodesWithLatency, RkvNode{Name: lnodes[0].String(), Latency: sahm.LatencyMap[lnodes[0].String()]})
+
 	}
-	remoteNodes := []rkvNode{}
-	rnodes := h.RemoteHasing.LocateNodes(key, h.RemoteCount)
-	if len(rnodes) != h.RemoteCount {
-		return nil, nil, fmt.Errorf("failed to get %d remote nodes. The return number is %d", h.RemoteCount, len(rnodes))
+	sort.Slice(nodesWithLatency, func(i, j int) bool {
+		return nodesWithLatency[i].Latency < nodesWithLatency[j].Latency
+	})
+
+	for _, node := range nodesWithLatency {
+		localNodes = append(localNodes, node)
+	}
+	return localNodes, nil
+}
+
+func (sahm SyncByZoneAsyncHashingManager) GetAsyncNodes(key []byte) ([]Node, error) {
+	if sahm.RemoteCount < 1 {
+		return make([]Node, 0), nil
+	}
+	rnodes := sahm.RemoteHasing.LocateNodes(key, sahm.RemoteCount)
+	if len(rnodes) != sahm.RemoteCount {
+		return nil, fmt.Errorf("failed to get %d remote nodes. The return number is %d", sahm.RemoteCount, len(rnodes))
+	}
+	return rnodes, nil
+}
+
+func (sahm SyncByZoneAsyncHashingManager) GetNodes(key []byte) ([]string, error) {
+	syncNodes, err := sahm.GetSyncNodes(key)
+	res := make([]string, 0)
+	if err != nil {
+		klog.Errorf("failed to get all the sync nodes: %v", err)
+		return res, err
+	}
+	syncNodesString := strings.Join(convertNodeArrToStringArr(syncNodes), ",")
+	asyncNodes, err := sahm.GetAsyncNodes(key)
+	if err != nil {
+		klog.Errorf("failed to get all the async nodes: %v", err)
+		return res, err
 	}
-	for _, rn := range rnodes {
-		remoteNodes = append(remoteNodes, rkvNode(rn.String()))
+	asyncNodesString := strings.Join(convertNodeArrToStringArr(asyncNodes), ",")
+	return []string{syncNodesString, asyncNodesString}, nil
+}
+
+func Factory(hashingType constants.ConsistentHashingType) ConsistentHashing {
```

Comment:
the name "Factory" can be more specific. Like "ConsistentHashingFactory"

Reply:
every package has a factory to create instances.

Reply:
got it.